compiler_opt/rl/regalloc/config.py

```diff
@@ -14,23 +14,68 @@
 # limitations under the License.
 """Register allocation training config."""
 
+from numpy import dtype, maximum
 import gin
 import tensorflow as tf
 from tf_agents.specs import tensor_spec
 from tf_agents.trajectories import time_step
 from compiler_opt.rl import feature_ops
 
+from absl import logging
 
 def get_num_registers():
   return 33
 
+def get_opcode_count():
+  return 17716
+
+def get_num_instructions():
+  return 300
+
+class process_instruction_features(tf.keras.Model):
+
+  def __init__(self, opcode_count, register_count, instruction_count, embedding_dimensions=16):
+    super().__init__()
+    self.opcode_count = opcode_count
+    self.register_count = register_count
+    self.instruction_count = instruction_count
+    self.embedding_dimensions = embedding_dimensions
+    self.embedding_layer = tf.keras.layers.Embedding(self.opcode_count,
+                                                     self.embedding_dimensions,
+                                                     input_length=self.instruction_count)
+
+  def get_config(self):
+    return {
+      "opcode_count": self.opcode_count,
+      "register_count": self.register_count,
+      "instruction_count": self.instruction_count,
+      "embedding_dimensions": self.embedding_dimensions
+    }
+
+  @classmethod
+  def from_config(cls, config):
+    return cls(**config)
+
+  def call(self, inputs):
+    # The first row is the instructions, and all rows afterwards compose the binary
+    # mapping matrix
+    instruction_opcodes = tf.slice(inputs, begin=[0, 0, 0], size=[-1, 1, self.instruction_count])
+    instruction_opcodes = tf.reshape(instruction_opcodes, [-1, self.instruction_count])
+    binary_mapping_matrix = tf.slice(inputs, begin=[0, 1, 0], size=[-1, self.register_count, self.instruction_count])
+    binary_mapping_matrix_casted = tf.cast(binary_mapping_matrix, tf.float32)
+    instruction_embeddings = self.embedding_layer(instruction_opcodes)
+
+    matrix_product = tf.linalg.matmul(binary_mapping_matrix_casted, instruction_embeddings)
```

Comment:
this is the bag-of-word aggregation I mentioned

Reply:
Yes. Just grabbing embeddings and then doing the matrix multiplication. There are a couple other options that I want to look at for the implementation of the bag of words aggregation that would in theory be more efficient than matrix multiplication. I'm going to try and get some code written up for that today.